### PR TITLE
Try to fix #76

### LINKED ISSUE #76: Race in the zip handling code
We've been seeing some weird behaviors when installing d3.js with jspm (`jspm install d3=github:mbostock/d3@3.4.10`). The main js file (github/mbostock@3.4.10/d3.js) gets corrupted (truncated) in some cases.
We've tried to create a reproducible test case without much success. We've tried with a different d3.js version but the problem also appears

After some debugging, we think the bug is in this [piece of code](https://github.com/jspm/github/blob/master/github.js#L569-L595).
The promise is resolved as soon as yauzl emits the "close" event. But from our understanding, the close signal is emitted once "all streams created from openReadStream() have emitted their end events.", but without ensuring that all the data has actually been processed (unzipped) and wrote to the output file.
We've managed to fix the bug by rewriting the code to only resolve the promise once all the file had finished writing (.on("close") for each write stream).

### PR DESCRIPTION
**Pull request by Gu1**
This code might not be optimal, but it'll give you an idea...
Still couldn't find a way to reproduce this bug. 
Don't know why it only happens with d3.js either.
I tried creating a release on a github project with a zip file attached with multiple big .js files inside. I then tried to install it through jspm. I confirmed the yauzl code path was hit, but the bug did not happen (extracted files were not corrupted..), so I'm still baffled.

### CODE REVIEW

File: github.js [hunk 1]
```diff
@@ -5,6 +5,7 @@ var rimraf = require('rimraf');
 var request = require('request');
 var expandTilde = require('expand-tilde');
 
+var rsvp = require('rsvp');
```
Comment (guybedford) [inline github.js]: I'd prefer to avoid the use of the non-ES6 promise API.

File: github.js [hunk 2]
```diff
@@ -584,14 +585,26 @@ GithubLocation.prototype = {
...
+                            });
+                            readStream.pipe(p);
+                        }));
                       });
                     });
                   });
-
-                  zipFile.on('close', resolve);
+                  zipFile.on('close', function() {
+                      rsvp.all(files).then(function() {
```
Comment (guybedford) [inline github.js]: `Promise.all(files).then(function() { ...` would be the ES promise equivalent here.

**Comment by guybedford** [comment]
This is excellent, thank you so much for the PR. let me know if the change is possible otherwise I can go ahead and merge manually.

**Comment by Gu1** [comment]
I just updated the pull request, following your recommendations.
Sorry, I thought that rsvp did not implement `Promise.all` and that you had to go through rsvp.all (I only read the example on their [README](https://github.com/tildeio/rsvp.js/#arrays-of-promises)).

The only problems I see with my patch is the fact that it creates one Promise object per file in the zip archive.. Don't know if it's a good thing performance/memory wise, especially when extracting a zip file with many many files inside ?..

**Comment by guybedford** [comment]
The good news is that this goes away in jspm 0.17 anyway, so I'm happy to merge. Thanks again for this.

**Pull request merged**